{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOXpcxeGPkyWi3/Ahns1TEq"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')\n","\n","import sys\n","sys.path.append('/content/drive/My Drive/Colab Notebooks/')\n","!cp -r \"/content/drive/My Drive/Colab Notebooks/unsw_processing.ipynb\" '/content/'\n","\n","!pip install import_ipynb\n","import import_ipynb\n","\n","from unsw_processing import unsw_encoding\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"sDt_fvEEKSsA","executionInfo":{"status":"ok","timestamp":1669201564351,"user_tz":0,"elapsed":7913,"user":{"displayName":"Diogo Beijinha","userId":"06164615948630766952"}},"outputId":"08b007d6-13bd-49ca-ece6-89aa5084ce07"},"execution_count":5,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n","Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Requirement already satisfied: import_ipynb in /usr/local/lib/python3.7/dist-packages (0.1.4)\n","Requirement already satisfied: IPython in /usr/local/lib/python3.7/dist-packages (from import_ipynb) (7.9.0)\n","Requirement already satisfied: nbformat in /usr/local/lib/python3.7/dist-packages (from import_ipynb) (5.7.0)\n","Requirement already satisfied: pygments in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (2.6.1)\n","Requirement already satisfied: pexpect in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (4.8.0)\n","Requirement already satisfied: jedi>=0.10 in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (0.18.2)\n","Requirement already satisfied: prompt-toolkit<2.1.0,>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (2.0.10)\n","Requirement already satisfied: backcall in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (0.2.0)\n","Requirement already satisfied: pickleshare in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (0.7.5)\n","Requirement already satisfied: traitlets>=4.2 in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (5.1.1)\n","Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (57.4.0)\n","Requirement already satisfied: decorator in /usr/local/lib/python3.7/dist-packages (from IPython->import_ipynb) (4.4.2)\n","Requirement already satisfied: parso<0.9.0,>=0.8.0 in /usr/local/lib/python3.7/dist-packages (from jedi>=0.10->IPython->import_ipynb) (0.8.3)\n","Requirement already satisfied: six>=1.9.0 in /usr/local/lib/python3.7/dist-packages (from prompt-toolkit<2.1.0,>=2.0.0->IPython->import_ipynb) (1.15.0)\n","Requirement already satisfied: wcwidth in /usr/local/lib/python3.7/dist-packages (from prompt-toolkit<2.1.0,>=2.0.0->IPython->import_ipynb) (0.2.5)\n","Requirement already satisfied: fastjsonschema in /usr/local/lib/python3.7/dist-packages (from nbformat->import_ipynb) (2.16.2)\n","Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.7/dist-packages (from nbformat->import_ipynb) (4.3.3)\n","Requirement already satisfied: jupyter-core in /usr/local/lib/python3.7/dist-packages (from nbformat->import_ipynb) (4.11.2)\n","Requirement already satisfied: importlib-metadata>=3.6 in /usr/local/lib/python3.7/dist-packages (from nbformat->import_ipynb) (4.13.0)\n","Requirement already satisfied: typing-extensions>=3.6.4 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata>=3.6->nbformat->import_ipynb) (4.1.1)\n","Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata>=3.6->nbformat->import_ipynb) (3.10.0)\n","Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import_ipynb) (0.19.2)\n","Requirement already satisfied: attrs>=17.4.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import_ipynb) (22.1.0)\n","Requirement already satisfied: importlib-resources>=1.4.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import_ipynb) (5.10.0)\n","Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.7/dist-packages (from pexpect->IPython->import_ipynb) (0.7.0)\n"]}]},{"cell_type":"code","execution_count":6,"metadata":{"id":"2kNefF3--RUf","executionInfo":{"status":"ok","timestamp":1669201564353,"user_tz":0,"elapsed":28,"user":{"displayName":"Diogo Beijinha","userId":"06164615948630766952"}}},"outputs":[],"source":["import os\n","#os.environ[\"CUDA_DEVICE_ORDER\"]=\"PCI_BUS_ID\" #If the line below doesn't work, uncomment this line (make sure to comment the line below); it should help.\n","os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"-1\"\n","import numpy as np\n","import tensorflow\n","from keras.models import load_model\n","from sklearn.metrics import (confusion_matrix, roc_auc_score,\n","                             precision_score, auc)\n"]},{"cell_type":"code","source":["# ***** REFERENCES PARAMETERS *****\n","params = {'train_data': 494021, 'features_nb': 4,\n","          'batch_size': 1024, 'encoder': 'standarscaler',\n","          'dataset': 'kdd'}\n","\n","model_name = 'content/models/' + '494021_4_mse_nadam_sigmoid_1_128_1024' + \\\n","    '_0.2_CuDNNLSTM_standarscaler_1562685990.8704927st'\n"],"metadata":{"id":"HBEQWnPpJSSu","executionInfo":{"status":"ok","timestamp":1669201564355,"user_tz":0,"elapsed":26,"user":{"displayName":"Diogo Beijinha","userId":"06164615948630766952"}}},"execution_count":7,"outputs":[]},{"cell_type":"code","source":["\n","# Encode dataset and return : x_train, x_test, y_train, y_test\n","def load_data():\n","    if params['dataset'] == 'unsw':\n","        x_train, x_test, y_train, y_test = unsw_encoding(params)\n","\n","    # Reshape the inputs in the accepted model format\n","    x_train = np.array(x_train).reshape([-1, x_train.shape[1], 1])\n","    x_test = np.array(x_test).reshape([-1, x_test.shape[1], 1])\n","    return x_train, x_test, y_train, y_test\n","\n","\n","# Print information on the results\n","def print_results(params, model, x_train, x_test, y_train, y_test):\n","    print('Val loss and acc:')\n","    print(model.evaluate(x_test, y_test, params['batch_size']))\n","\n","    y_pred = model.predict(x_test, params['batch_size'])\n","\n","    print('\\nConfusion Matrix:')\n","    conf_matrix = confusion_matrix(y_test.argmax(axis=1),\n","                                   y_pred.argmax(axis=1))\n","    print(conf_matrix)\n","\n","    FP = conf_matrix.sum(axis=0) - np.diag(conf_matrix)  # False Positive\n","    FN = conf_matrix.sum(axis=1) - np.diag(conf_matrix)  # False Negative\n","    TP = np.diag(conf_matrix)  # True Positive\n","    TN = conf_matrix.sum() - (FP + FN + FP)  # True Negative\n","\n","    print('\\nTPR:')  # True Positive Rate\n","    # Portion of positive instances correctly predicted positive\n","    print(TP / (TP + FN))\n","\n","    print('\\nFPR:')  # False Positive Rate\n","    # Portion of negative instances wrongly predicted positive\n","    print(FP / (FP + TN))\n","\n","    # Cost Matrix as presented in Staudemeyer article\n","    cost_matrix = [[0, 1, 2, 2, 2],\n","                   [1, 0, 2, 2, 2],\n","                   [2, 1, 0, 2, 2],\n","                   [4, 2, 2, 0, 2],\n","                   [4, 2, 2, 2, 0]]\n","\n","    tmp_matrix = np.zeros((5, 5))\n","\n","    for i in range(5):\n","        for j in range(5):\n","            tmp_matrix[i][j] = conf_matrix[i][j] * cost_matrix[i][j]\n","\n","    # The average cost is (total cost / total number of classifications)\n","    print('\\nCost:')\n","    print(tmp_matrix.sum()/conf_matrix.sum())\n","\n","    print('\\nAUC:')  # Average Under Curve\n","    print(roc_auc_score(y_true=y_test, y_score=y_pred, average=None))\n","\n","    print('\\nPrecision:')  # Probability an instance gets correctly predicted\n","\n","    print(precision_score(y_true=y_test.argmax(axis=1),\n","                          y_pred=y_pred.argmax(axis=1), average=None))\n","\n","\n","if __name__ == \"__main__\":\n","    # Allows tensorflow to run multiple sessions and learning simultaneously\n","    # Comment the 3 following lines if causing issues\n","    # config = tensorflow.ConfigProto()\n","    # config.gpu_options.allow_growth = True\n","    # sess = tensorflow.Session(config=config)\n","\n","    model = load_model(model_name)\n","    model.summary()\n","\n","    x_train, x_test, y_train, y_test = load_data()\n","    print_results(params, model, x_train, x_test, y_train, y_test)"],"metadata":{"id":"H63qtfvRJaYy","colab":{"base_uri":"https://localhost:8080/","height":380},"executionInfo":{"status":"error","timestamp":1669201564356,"user_tz":0,"elapsed":26,"user":{"displayName":"Diogo Beijinha","userId":"06164615948630766952"}},"outputId":"b5bc5b74-32a3-407e-c3de-7f114d74e0e0"},"execution_count":8,"outputs":[{"output_type":"error","ename":"OSError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)","\u001b[0;32m<ipython-input-8-3e77aa8f3ef6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     68\u001b[0m     \u001b[0;31m# sess = tensorflow.Session(config=config)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     69\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 70\u001b[0;31m     \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mload_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     71\u001b[0m     \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msummary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     65\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# pylint: disable=broad-except\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m       \u001b[0mfiltered_tb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_process_traceback_frames\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__traceback__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 67\u001b[0;31m       \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     68\u001b[0m     \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     69\u001b[0m       \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.7/dist-packages/keras/saving/save.py\u001b[0m in \u001b[0;36mload_model\u001b[0;34m(filepath, custom_objects, compile, options)\u001b[0m\n\u001b[1;32m    204\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m           \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 206\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mIOError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'No file or directory found at {filepath_str}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    207\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m           \u001b[0;32mif\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mOSError\u001b[0m: No file or directory found at content/models/494021_4_mse_nadam_sigmoid_1_128_1024_0.2_CuDNNLSTM_standarscaler_1562685990.8704927st"]}]}]}